File: test/fixengine/messages/FixMessageParserSpec.java
/*
 * Copyright 2010 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package fixengine.messages;

import java.nio.ByteBuffer;

import org.junit.Assert;
import org.junit.Test;

import silvertip.GarbledMessageException;
import silvertip.Message;
import silvertip.PartialMessageException;

public class FixMessageParserSpec {
  private static final int RECEIVE_BUFFER_SIZE = 4096;
  private static final char DELIMITER = '\001';

  private final ByteBuffer rxBuffer = ByteBuffer.allocate(RECEIVE_BUFFER_SIZE);
  private FixMessageParser parser = new FixMessageParser();

  @Test(expected = PartialMessageException.class)
  public void empty() throws Exception {
    parse("");
  }

  @Test(expected = PartialMessageException.class)
  public void partialHeader() throws Exception {
    String partialHeader = "8=FIX.4.2" + DELIMITER + "9=153";

    parse(partialHeader);
  }

  @Test public void firstFieldIsNotBeginString() throws Exception {
    String garbled = "X=FIX.4.2" + DELIMITER + "9=5" + DELIMITER;

    assertGarbledMessage(garbled, "Expected tag not found: BeginString(8)");
  }

  @Test public void emptyBeginString() throws Exception {
    String garbled = "8=" + DELIMITER + "9=5" + DELIMITER;

    assertGarbledMessage(garbled, "BeginString(8) is empty");
  }

  @Test public void invalidBeginString() throws Exception {
    String garbled = "8=XIF.10.1" + DELIMITER + "9=5" + DELIMITER;

    assertGarbledMessage(garbled, "BeginString(8) is invalid, expected format FIX.m.n: XIF.10.1");
  }

  @Test public void secondFieldIsNotBodyLength() throws Exception {
    String garbled = "8=FIX.4.2" + DELIMITER + "X=5" + DELIMITER;

    assertGarbledMessage(garbled, "Expected tag not found: BodyLength(9)");
  }

  @Test public void secondFieldIsNotBodyLengthAndIsMissing() throws Exception {
    String garbled = "8=FIX.4.2" + DELIMITER;

    assertGarbledMessage(garbled, "Expected tag not found: BodyLength(9)");
  }

  @Test public void emptyBodyLength() throws Exception {
    String garbled = "8=FIX.4.2" + DELIMITER + "9=" + DELIMITER + "108=XXX" + DELIMITER;

    assertGarbledMessage(garbled, "BodyLength(9) is empty");
  }

  @Test public void invalidBodyLengthFormat() throws Exception {
    String garbled = "8=FIX.4.2" + DELIMITER + "9=XXX" + DELIMITER + "108=XXX" + DELIMITER;

    assertGarbledMessage(garbled, "BodyLength(9) has invalid format: XXX");
  }

  @Test public void bodyLengthDoesNotPointToStandardTrailer() throws Exception {
    String header = "8=FIX.4.2" + DELIMITER + "9=5" + DELIMITER;
    String payload = "35=E" + DELIMITER;
    String trailer = "10=XXX" + DELIMITER;
    String messages = header + payload + header + payload + trailer;
    String garbled = header + payload;

    assertGarbledMessage(messages, garbled, "Expected tag not found: CheckSum(10)");
  }

  @Test public void thirdFieldIsNotMsgType() throws Exception {
    String garbled = "8=FIX.4.2" + DELIMITER + "9=5" + DELIMITER + "X=A" + DELIMITER;

    assertGarbledMessage(garbled, "Expected tag not found: MsgType(35)");
  }

  @Test public void thirdFieldIsNotMsgTypeAndIsMissing() throws Exception {
    String garbled = "8=FIX.4.2" + DELIMITER + "9=5" + DELIMITER;

    assertGarbledMessage(garbled, "Expected tag not found: MsgType(35)");
  }

  @Test public void emptyMsgType() throws Exception {
    String garbled = "8=FIX.4.2" + DELIMITER + "9=5" + DELIMITER + "35=" + DELIMITER + "108=XXX" + DELIMITER;

    assertGarbledMessage(garbled, "MsgType(35) is empty");
  }

  @Test public void missingCheckSum() throws Exception {
    String header = "8=FIX.4.2" + DELIMITER + "9=5" + DELIMITER;
    String payload = "35=E" + DELIMITER;
    String trailer = "11=XXX" + DELIMITER;
    String message = header + payload + trailer;

    assertGarbledMessage(message, "Expected tag not found: CheckSum(10)");
  }

  @Test public void emptyCheckSum() throws Exception {
    String header = "8=FIX.4.2" + DELIMITER + "9=5" + DELIMITER;
    String payload = "35=E" + DELIMITER;
    String trailer = "10=" + DELIMITER;
    String message = header + payload + trailer;

    assertGarbledMessage(message, "CheckSum(10) is empty");
  }

  @Test public void invalidCheckSumFormat() throws Exception {
    String garbled = "8=FIX.4.2" + DELIMITER + "9=5" + DELIMITER + "35=E" + DELIMITER + "10=XYZ" + DELIMITER;

    assertGarbledMessage(garbled, "CheckSum(10) has invalid format: XYZ");
  }

  @Test public void invalidCheckSumLength() throws Exception {
    String garbled = "8=FIX.4.2" + DELIMITER + "9=5" + DELIMITER + "35=E" + DELIMITER + "10=12" + DELIMITER;

    assertGarbledMessage(garbled, "CheckSum(10) has invalid length: 12");
  }

  @Test public void invalidCheckSum() throws Exception {
    String garbled = "8=FIX.4.2" + DELIMITER + "9=5" + DELIMITER + "35=E" + DELIMITER + "10=123" + DELIMITER;

    assertGarbledMessage(garbled, "CheckSum(10) mismatch, expected=182, actual=123");
  }

  @Test public void garbledMessageEndingWith8SOH() throws Exception {
    String garbled = "8=FIX.4.2" + DELIMITER + "9=5" + DELIMITER + "8" + DELIMITER;

    assertGarbledMessage(garbled, "Expected tag not found: MsgType(35)");
  }

  @Test public void garbledMessageEndingWith8SOHSOH() throws Exception {
    String garbled = "8=FIX.4.2" + DELIMITER + "9=5" + DELIMITER + "8" + DELIMITER + DELIMITER;

    assertGarbledMessage(garbled, "Expected tag not found: MsgType(35)");
  }

  @Test(expected = PartialMessageException.class)
  public void partialMessage() throws Exception {
    String header = "8=FIX.4.2" + DELIMITER + "9=153" + DELIMITER + "";

    parse(header);
  }

  @Test
  public void fullMessage() throws Exception {
    String header = "8=FIX.4.2" + DELIMITER + "9=153" + DELIMITER + "";
    String payload = "35=E" + DELIMITER + "49=INST" + DELIMITER + "56=BROK" + DELIMITER + "52=20050908-15:51:22"
        + DELIMITER + "34=200" + DELIMITER + "66=14" + DELIMITER + "394=1" + DELIMITER + "68=2" + DELIMITER + "73=2"
        + DELIMITER + "11=order-1" + DELIMITER + "67=1" + DELIMITER + "55=IBM" + DELIMITER + "54=2" + DELIMITER
        + "38=2000" + DELIMITER + "40=1" + DELIMITER + "11=order-2" + DELIMITER + "67=2" + DELIMITER + "55=AOL"
        + DELIMITER + "54=2" + DELIMITER + "38=1000" + DELIMITER + "40=1" + DELIMITER + "";
    String trailer = "10=020" + DELIMITER;

    Message message = parse(header + payload + trailer);
    Assert.assertEquals(header + payload + trailer, message.toString());
  }

  private void assertGarbledMessage(String garbledMessage, String expectedReason) throws Exception {
    String valid = "8=FIX.4.2" + DELIMITER + "9=5" + DELIMITER + "35=E" + DELIMITER + "10=191" + DELIMITER;
    assertGarbledMessage(garbledMessage + valid, garbledMessage, expectedReason);
  }

  private void assertGarbledMessage(String messages, String expectedGarbledMessage, String expectedReason) throws Exception {
    try {
      parse(messages);
      Assert.fail("GarbledMessageException was not thrown for message: " + expectedGarbledMessage);
    } catch (GarbledMessageException e) {
      Assert.assertEquals(expectedGarbledMessage, new String(e.getMessageData()));
      Assert.assertEquals(expectedReason, e.getMessage());
    }
  }

  private Message parse(String message) throws PartialMessageException, GarbledMessageException {
    rxBuffer.put(message.getBytes());
    rxBuffer.flip();
    return parse(rxBuffer);
  }

  private Message parse(ByteBuffer buffer) throws PartialMessageException, GarbledMessageException {
    buffer.mark();
    return parser.parse(buffer);
  }
}

File: src/test/scala/stirling/fix/messages/FixMessageParserSpec.scala
/*
 * Copyright 2010 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package stirling.fix.messages

import java.nio.ByteBuffer
import org.scalatest.{Assertions, WordSpec}
import org.scalatest.matchers.MustMatchers
import silvertip.{GarbledMessageException, PartialMessageException}

class FixMessageParserSpec extends WordSpec with MustMatchers with FixMessageParserFixtures {
  "FixMessageParser" should {
    val parser = new FixMessageParser()
    "throw an exception on an empty buffer" in {
      intercept[PartialMessageException] {
        parser.parse("")
      }
    }
    "throw an exception on a partial header" in {
      intercept[PartialMessageException] {
        parser.parse("8=FIX.4.2" + delimiter + "9=153")
      }
    }
    "throw an exception if the first field is not BeginString(8)" in {
      val message = "X=FIX.4.2" + delimiter + "9=5" + delimiter
      assertGarbledMessage(parser, message, "Expected tag not found: BeginString(8)")
    }
    "throw an exception on an empty BeginString(8)" in {
      val message = "8=" + delimiter + "9=5" + delimiter
      assertGarbledMessage(parser, message, "BeginString(8) is empty")
    }
    "throw an exception on an invalid BeginString(8)" in {
      val message = "8=XIF.10.1" + delimiter + "9=5" + delimiter
      assertGarbledMessage(parser, message, "BeginString(8) is invalid, expected format FIX.m.n: XIF.10.1")
    }
    "throw an exception if second field is missing" in {
      val message = "8=FIX.4.2" + delimiter
      assertGarbledMessage(parser, message, "Expected tag not found: BodyLength(9)")
    }
    "throw an exception if second field is not BodyLength(9)" in {
      val message = "8=FIX.4.2" + delimiter + "X=5" + delimiter
      assertGarbledMessage(parser, message, "Expected tag not found: BodyLength(9)")
    }
    "throw an exception on an empty BodyLength(9)" in {
      val message = "8=FIX.4.2" + delimiter + "9=" + delimiter + "108=XXX" + delimiter
      assertGarbledMessage(parser, message, "BodyLength(9) is empty")
    }
    "throw an exception on an invalid format in BodyLength(9)" in {
      val message = "8=FIX.4.2" + delimiter + "9=XXX" + delimiter + "108=XXX" + delimiter
      assertGarbledMessage(parser, message, "BodyLength(9) has invalid format: XXX")
    }
    "throw an exception if BodyLength(9) does not point to standard trailer" in {
      val header = "8=FIX.4.2" + delimiter + "9=5" + delimiter
      val payload = "35=E" + delimiter
      val trailer = "10=XXX" + delimiter
      val messages = header + payload + header + payload + trailer
      val garbled = header + payload
      assertGarbledMessage(parser, messages, garbled, "Expected tag not found: CheckSum(10)")
    }
    "throw an exception if the third field is not MsgType(35)" in {
      val message = "8=FIX.4.2" + delimiter + "9=5" + delimiter + "X=A" + delimiter
      assertGarbledMessage(parser, message, "Expected tag not found: MsgType(35)")
    }
    "throw an exception if the third field is missing" in {
      val message = "8=FIX.4.2" + delimiter + "9=5" + delimiter
      assertGarbledMessage(parser, message, "Expected tag not found: MsgType(35)")
    }
    "throw an exception on an empty MsgType(35)" in {
      val message = "8=FIX.4.2" + delimiter + "9=5" + delimiter + "35=" + delimiter + "108=XXX" + delimiter
      assertGarbledMessage(parser, message, "MsgType(35) is empty")
    }
    "throw an exception if CheckSum(10) is missing" in {
      val header = "8=FIX.4.2" + delimiter + "9=5" + delimiter
      val payload = "35=E" + delimiter
      val trailer = "11=XXX" + delimiter
      val message = header + payload + trailer
      assertGarbledMessage(parser, message, "Expected tag not found: CheckSum(10)")
    }
    "throw an exception on an empty CheckSum(10)" in {
      val header = "8=FIX.4.2" + delimiter + "9=5" + delimiter
      val payload = "35=E" + delimiter
      val trailer = "10=" + delimiter
      val message = header + payload + trailer
      assertGarbledMessage(parser, message, "CheckSum(10) is empty")
    }
    "throw an exception on an invalid CheckSum(10)" in {
      val message = "8=FIX.4.2" + delimiter + "9=5" + delimiter + "35=E" + delimiter + "10=XYZ" + delimiter
      assertGarbledMessage(parser, message, "CheckSum(10) has invalid format: XYZ")
    }
    "throw an exception on an invalid length on CheckSum(10)" in {
      val message = "8=FIX.4.2" + delimiter + "9=5" + delimiter + "35=E" + delimiter + "10=12" + delimiter
      assertGarbledMessage(parser, message, "CheckSum(10) has invalid length: 12")
    }
    "throw an exception on invalid CheckSum(10)" in {
      val message = "8=FIX.4.2" + delimiter + "9=5" + delimiter + "35=E" + delimiter + "10=123" + delimiter
      assertGarbledMessage(parser, message, "CheckSum(10) mismatch, expected=182, actual=123")
    }
    "throw an exception on a garbled message ending with 8SOH" in {
      val message = "8=FIX.4.2" + delimiter + "9=5" + delimiter + "8" + delimiter
      assertGarbledMessage(parser, message, "Expected tag not found: MsgType(35)")
    }
    "throw an exception on a garbled message ending with 8SOHSOH" in {
      val message = "8=FIX.4.2" + delimiter + "9=5" + delimiter + "8" + delimiter + delimiter
      assertGarbledMessage(parser, message, "Expected tag not found: MsgType(35)")
    }
    "throw an exception on a partial message" in {
      intercept[PartialMessageException] {
        parser.parse("8=FIX.4.2" + delimiter + "9=153" + delimiter)
      }
    }
    "accept a valid message" in {
      val header = "8=FIX.4.2" + delimiter + "9=153" + delimiter
      val payload = "35=E" + delimiter + "49=INST" + delimiter + "56=BROK" + delimiter + "52=20050908-15:51:22" +
        delimiter + "34=200" + delimiter + "66=14" + delimiter + "394=1" + delimiter + "68=2" + delimiter + "73=2" +
        delimiter + "11=order-1" + delimiter + "67=1" + delimiter + "55=IBM" + delimiter + "54=2" + delimiter +
        "38=2000" + delimiter + "40=1" + delimiter + "11=order-2" + delimiter + "67=2" + delimiter + "55=AOL" +
        delimiter + "54=2" + delimiter + "38=1000" + delimiter + "40=1" + delimiter
      val trailer = "10=020" + delimiter
      val message = parser.parse(header + payload + trailer)
      message.toString must equal(header + payload + trailer)
    }
  }
}

trait FixMessageParserFixtures extends Assertions with MustMatchers {
  def delimiter = '\001'
  def bufferSize = 4096
  implicit def stringToByteBuffer(string: String): ByteBuffer = {
    val buffer = ByteBuffer.allocate(bufferSize)
    buffer.put(string.getBytes)
    buffer.flip
    buffer.mark
    buffer
  }
  def assertGarbledMessage(parser: FixMessageParser, garbledMessage: String, expectedReason: String) {
    val validMessage = "8=FIX.4.2" + delimiter + "9=5" + delimiter + "35=E" + delimiter + "10=191" + delimiter
    assertGarbledMessage(parser, garbledMessage + validMessage, garbledMessage, expectedReason)
  }
  def assertGarbledMessage(parser: FixMessageParser, messages: String, expectedGarbledMessage: String, expectedReason: String) {
    val exception = intercept[GarbledMessageException] {
      parser.parse(messages)
    }
    new String(exception.getMessageData) must equal(expectedGarbledMessage)
    exception.getMessage must equal(expectedReason)
  }
}
